src/state_circuit/state.rs

```diff
@@ -0,0 +1,896 @@
+            || "address table with zero",
+            |mut region| {
+                for idx in 0..=ADDRESS_MAX {
+                    region.assign_fixed(
+                        || "address table with zero",
+                        self.address_table_zero,
+                        idx,
+                        || Ok(F::from_u64(idx as u64)),
+                    )?;
+                }
+                Ok(())
+            },
+        )
+    }
+
+    /// Assign cells.
+    pub(crate) fn assign(
+        &self,
+        mut layouter: impl Layouter<F>,
+        ops: Vec<Op<F>>,
+    ) -> Result<Vec<BusMapping<F>>, Error> {
+        let mut bus_mappings: Vec<BusMapping<F>> = Vec::new();
+
+        let chip = IsZeroChip::construct(self.is_zero.clone());
+        let monotone_chip =
+            MonotoneChip::<F, ADDRESS_MAX, ADDRESS_INCR, ADDRESS_DIFF_STRICT>::construct(
+                self.mono_incr.clone(),
+            );
+        monotone_chip.load(&mut layouter)?;
+
+        layouter.assign_region(
+            || "State operations",
+            |mut region| {
+                let mut offset = 0;
+
+                for (index, op) in ops.iter().enumerate() {
+                    let address = op.address;
+                    let mut address_prev = ops.first().unwrap().address;
+                    if index > 0 {
+                        address_prev = ops[index - 1].address;
+                    }
+
+                    if index == 0 {
+                        self.q_first.enable(&mut region, offset)?;
+                    } else {
+                        self.q_not_first.enable(&mut region, offset)?;
+                    }
+                    self.init(&mut region, offset, address)?;
+                    chip.assign(&mut region, offset, Some(address.0 - address_prev.0))?;
+
+                    // Increase offset by 1 after initialising.
+                    offset += 1;
+
+                    for global_counter in op.global_counters.iter() {
+                        self.q_not_first.enable(&mut region, offset)?;
+                        let bus_mapping =
+                            self.assign_per_counter(&mut region, offset, address, global_counter)?;
+                        chip.assign(&mut region, offset, Some(F::zero()))?;
+
+                        offset += 1;
+
+                        // TODO: order by global_counter
+                        bus_mappings.push(bus_mapping);
+                    }
+                }
+
+                let last_address = ops.last().unwrap().address;
+                let last_gc = ops.last().unwrap().global_counters.last().unwrap();
+                let mut last_counter = last_gc.as_ref().unwrap().global_counter().0;
+
+                // We pad all remaining memory rows to avoid the check at the first unused row.
+                // Without padding, (address_cur - address_prev) would not be zero at the first unused row
+                // and some checks would be triggered.
```

Comment:
TODO: We eventually want to exclude the padding rows from the bus mapping lookup.